### PR TITLE
feat(u): BigInteger

### PR DESCRIPTION
**Pull request by snowypowers**
BigInteger using bn.js internally with utility methods for conversion around two's complement.

To add +,-,*,/,% operations.

### CODE REVIEW

**Comment by ixje** [comment]
Here are a couple of edge cases that don't work as expected in the "from 2 complement to bigint". 
| hex string       | Expected             | Actual |
|------------------|----------------------|--------|
| 8000             | 128                  | -32768 |
| FF00             | 255                  | -255   |
| 0004             | 1024                 | 4      |
| 0000000000000080 | -9223372036854775808 | 128    |
| FFFFFFFFFFFFFF7F | 9223372036854775807  | -129   |

This is mainly because C#'s BigInteger is not standard 2-complement. To quote their description (see the "remarks" section: [here](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.biginteger.tobytearray?view=netcore-3.1))
> Negative values are written to the array using two's complement representation **in the most compact form possible**. For example, -1 is represented as a single byte whose value is 0xFF instead of as an array with multiple elements, such as 0xFF, 0xFF or 0xFF, 0xFF, 0xFF, 0xFF.
> 
> Because two's complement representation always interprets the highest-order bit of the last byte in the array (the byte at position Array.Length- 1) as the sign bit, the method returns a byte array with an extra element whose value is zero to disambiguate positive values that could otherwise be interpreted as having their sign bits set. For example, the value 120 or 0x78 is represented as a single-byte array: 0x78. However, 128, or 0x80, is represented as a two-byte array: 0x80, 0x00.

Note that from BigInteger to a hex string will have similar issues. I don't think we can ignore these edge cases; for example a `BigInteger(128)` now translates to `80`, but on the C# side it will turn into `-128`.

**Comment by snowypowers** [comment]
the conversion from hexTwos to BigInteger should be no issue assuming that the user passes the hexstring as a whole correctly down. The issue would be the output (`toTwos()`) which was truncating the leading zeroes. I have fixed the output method to force it to output the exact amount of bytes that it requires so the leading zeroes stay.

- Add test cases 128, 255, 1024.
- Add math operations.
- Add more docstrings.

This is nearly done, just more test cases to cover fromHex and fromNumber but I do not foresee any complications with them.

File: packages/neon-core/src/u/BigInteger.ts [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    return new BigInteger(
+      new BN(cleanedInput.toBigEndian(), 16).fromTwos(
+        cleanedInput.byteLength * 8
+      )
+    );
+  }
+
+  /**
+   * Creates a BigInteger from a hexstring.
+   * @param hexstring - HexString class or a bigendian string.
```
Comment (ixje) [inline packages/neon-core/src/u/BigInteger.ts:42]:
```suggestion
   * @param hexstring - HexString class or a little endian string.
```
same as `fromTwos()`

**Review by ixje — CHANGES_REQUESTED** [review]
### OK
 - basic math works as intended
 - division uses the correct truncated division (not floored like Python did)
- modulo similarly uses correct truncated division
- possible to convert to correct hex strings
- `fromHex` parses input to big endian value by default (matching C#)

### Not OK
- documentation for `fromHex` and `fromTwos` mention the source string endianness in the wrong order. Commented at code locs
  
### Ideas
Assuming the BigInteger class will be public and is specifically intended for use with NEO, then I believe we can make it a tid bit friendlier 
- currently I can't seem to do arithmetic like `big * big` but must use `.mul()`, `.mod`, `.add()` etc

Some C# for reference
```C#
var b = new BigInteger(128);
Console.WriteLine(b.ToByteArray().ToHexString()); // 8000 (little endian by default)
Console.WriteLine(b.ToByteArray(isUnsigned: false, isBigEndian: false).ToHexString()); // 8000
Console.WriteLine(b.ToByteArray(isUnsigned: false, isBigEndian: true).ToHexString()); // 0080

Console.WriteLine(BigInteger.Parse("8000", NumberStyles.HexNumber)); // -32768 (from little endian)
Console.WriteLine(BigInteger.Parse("0080", NumberStyles.HexNumber)); // 128 (from big endian)
```
To make it less error prone for those that don't know the mapping to C# we could;
- remove the existing `toHex()` as it is wrong in the context of NEO's BigInteger
- rename `toTwos()` to `toHex()`. Optionally have a "as_little_endian/as_bigendian" like flag
- have `toTwos()` output as little endian by default (also requires docstring update) instead of big endian. This matches C#'s `BigInteger.ToByteArray()` with no arguments. 

   I have to say this one is a bit of a shit show on NEO's side where they mix big and little endian depending on where it is used. If I'm not mistaken in `neon-js` this class will be used for outputing and parsing virtual machine related scripts. In the VM the BigInteger class is used in the default little endian mode where the following holds (Note: Integer internally builds on BigInteger)
   ```C#
   var i = new Neo.VM.Types.Integer(128); 
   Console.WriteLine(i.GetSpan().ToHexString()); // 8000
   
   byte[] data = new byte[] {0x80, 0x00};
   var b_data = new BigInteger(data);
   Console.WriteLine(b_data); // 128
   ```

File: packages/neon-core/src/u/BigInteger.ts [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+/**
+ * This is a BigInteger representation, providing several helper methods to convert between different formats as well as basic math operations.
+ */
+export class BigInteger {
+  /** Internal holding value. */
+  #value: BN;
+
+  /**
+   * Creates a BigInteger from a twos complement hexstring.
+   * @param hexstring - HexString class or a bigendian string.
```
[the remark this answers is not in the document]
Comment (snowypowers) [reply packages/neon-core/src/u/BigInteger.ts:14]: This is intended. The internal model for `neon-js` attempts to implement `always big endian`.

`HexString` is another utility class implemented within `neon-js`. Treat it as a hexstring holder that is aware of endianness.

Please look at the examples just a line below. It should demonstrate the usage. 

We can make it clearer by updating it to `HexString class or a string. Indicate endianness of the string with the 2nd boolean parameter.`

[the remark this answers is not in the document]
Comment (ixje) [reply packages/neon-core/src/u/BigInteger.ts:14]: Nm this comment. Things are getting mixed up now, sorry.

File: packages/neon-core/src/u/BigInteger.ts [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+   * Creates a BigInteger from a hexstring.
+   * @param hexstring - HexString class or a bigendian string.
+   *
+   * @example
+   * const twoFiveFive = BigInteger.fromHex("ff");
+   * const oneTwoEight = BigIntger.fromHex("80");
+   * const oneTwoEightAgain = BigInteger.fromHex("0080");
+   *
+   * // Apply a second paramter when using little endian.
+   * const oneZeroTwoFour = BigInteger.fromHex("0040", true);
```
Comment (ixje) [inline packages/neon-core/src/u/BigInteger.ts:50]:
```suggestion
   * const oneSixThreeEightFour = BigInteger.fromHex("0040", true);
```

Here I get
```javascript
console.log(BigInteger.fromHex("0040", true).toString()); // 16384
```

**Comment by snowypowers** [comment]
> currently I can't seem to do arithmetic like big * big but must use .mul(), .mod, .add() etc
Unfortunately, operator overloads dont really work well in javascript. Not possible.

Not trying to match C# here. 

1. Renaming `toTwos` to `toHex` will require renaming `fromTwos` to `fromHex`. I rather not remove the functionality for now until im clear that there is no usage. If we ever discover we need the standard hex conversion, I rather have `toTwos` and `toHex` than `toHex` and `toNormalHex`(or whatever you can think of).
1. Lets add `toReverseTwos` that will match `toReverseHex` for little endian.
1. I would like to lean towards the default of bigendian for everything in `neon-js`. Makes it easier to work with. The HexString class was basically written to solve this issue of random endianness.

**Comment by ixje** [comment]
> 1. Renaming `toTwos` to `toHex` will require renaming `fromTwos` to `fromHex`. I rather not remove the functionality for now until im clear that there is no usage. If we ever discover we need the standard hex conversion, I rather have `toTwos` and `toHex` than `toHex` and `toNormalHex`(or whatever you can think of).

Perhaps it's an idea to add a note in the documentation that for use with NEO internals the `toTwos()/fromTwos()` is recommended? As an outsider the discussion we're having is pretty much impossible to be aware of unless you find this PR.

**Review by ixje — APPROVED** [review]

**Pull request merged**